source/MapDetailPanel.cpp

```diff
@@ -569,11 +570,29 @@ void MapDetailPanel::DrawKey()
 	}
 	else if(commodity == SHOW_GOVERNMENT)
 	{
+		unordered_map<string, const Government *> displayNames;
 		// Each system is colored by the government of the system. Only the
 		// four largest visible governments are labeled in the legend.
 		vector<pair<double, const Government *>> distances;
 		for(const auto &it : closeGovernments)
-			distances.emplace_back(it.second, it.first);
+		{
+			auto it2 = displayNames.find(it.first->GetName());
+			if(it2 != displayNames.end() && it2->second->GetColor() == it.first->GetColor())
```

Comment:
Should we check it2->second it-first for nullptr before de-referencing them? Some later code suggests that the Government-pointer can be nullptr (in which case the code starts using the colors and strings for Uninhabited).

Reply:
While the existing code does use `MapPanel::GovernmentColor(const Government *)`, which will check if hte pointer is null, it also then calls `GetName()` on that same potentially null pointer. If this was an issue, I think we would have seen it already?

```
		if(commodity == SHOW_GOVERNMENT && node.government && node.government->GetName() != "Uninhabited")
		{
			// For every government that is drawn, keep track of how close it
			// is to the center of the view. The four closest governments
			// will be displayed in the key.
			double distance = pos.Length();
			auto it = closeGovernments.find(node.government);
			if(it == closeGovernments.end())
				closeGovernments[node.government] = distance;
			else
				it->second = min(it->second, distance);
		}
	}
}
```
This code, in `MapPanel::DrawSystems()` seems to be the only place where this map is modified. It ensures that the Government pointer it is adding is not null.
`closeGovernments` is, however, a protected member of `MapPanel`, so it may be possible that any other child class may access it and insert null pointers in the future, so perhaps it is best to check?

Reply:
Agreed, sorry I didn't see that.